lib/fluent/root_agent.rb

```diff
@@ -231,13 +264,45 @@ def flush!
       flushing_threads.each{|t| t.join }
     end
 
-    def shutdown # Fluentd's shutdown sequence is stop, before_shutdown, shutdown, after_shutdown, close, terminate for plugins
+    def cancel_source_only!
+      unless @with_source_only
+        log.info "do nothing for canceling with-source-only because the current mode is not with-source-only."
+        return
+      end
+
+      log.info "cancel with-source-only mode and start the other plugins"
+      start
+
+      lifecycle_control_list[:input].each(&:event_emitter_cancel_source_only)
+
+      # Want to make sure that the source_only_router finishes all process before
+      # shutting down the agent.
+      # Strictly speaking, it would be necessary to have exclusive lock between
+      # EventRouter and the shutting down process of this agent.
+      # However, adding lock to EventRouter would worsen its performance, and
+      # the entire shutting down process does not care about it either.
+      # So, just sleep a few seconds here, though it may not be strictly safe.
+      sleep 5
```

Comment:
As commented, we cannot guarantee safety, no matter how large this value is.

[The entire shutting down process](https://github.com/fluent/fluentd/blob/0e96d5485912a099c0e445b62bbab9963c579601/lib/fluent/root_agent.rb#L298-L371) does not care about this (EventRouter must finish all process before filter/output plugins shutdowns.)
So, it would not actually cause problems, but I put `sleep` for a few seconds just to be safe.

Reply:
> If it doesn't cause problems, isn't it unnecessary?

I'm not sure.
I am only assuming that, based on existing implementations, such problems will basically not occur.
I expect some `sleep` would reduce the possibility to almost zero.

> I don't want to put in a magic number that I'll be stuck dealing with when I read it later.

I understand that. So, I wrote the comments there.
Is it insufficient? 

Since we cannot deny the possibility that this `sleep` may contribute to the stability of the operation, it is hard to say whether it is really better to remove it.

I understand that an implementation that does not require comments is better, but should we prioritize it over one that may contribute to stability of the operation?

Maybe we should verify this effect.

Reply:
Hmm, as commented in the codes,  it would be tough to solve the root cause.

> Strictly speaking, it would be necessary to have exclusive lock between
> EventRouter and the shutting down process of this agent.
> However, adding lock to EventRouter would worsen its performance

Reply:
> it would be necessary to have exclusive lock between EventRouter and the shutting down process.

I'm checking if this is true.

Reply:
I have checked this.
The conclusion is that ...

* We should `sleep` for 1 sec, at least.
* We should overwrite `handle_emits_error` of `SourceOnlyBufferAgent`.

Here are the details.

https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/event_router.rb#L114-L118

I can reproduce it by adding `sleep` to `EventRouter#emit_stream()`.

```diff
    def emit_stream(tag, es)
+     sleep 10
      match(tag).emit_events(tag, es)
```

Sending SIGWINCH will terminate the buffer.
Then, some crucial variables become `nil`, and it causes `nil` error.

https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/plugin/buffer.rb#L285-L292
https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/plugin/buffer.rb#L328-L330
https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/plugin/buffer.rb#L294-L296

The problem is that we can not confirm the error in the current implementation.
We should overwrite `handle_emits_error` of `SourceOnlyBufferAgent`.

https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/event_router.rb#L114-L123
https://github.com/fluent/fluentd/blob/eebc7e1da6e0916be2ccdfab79fe0cffef296d21/lib/fluent/agent.rb#L162-L166

Reply:
> We should sleep for 1 sec, at least.

`1s` would be enough because the router has only one output plugin when using source-only.
So, `EventRouter#emit_stream()` would immediately pass the data to the output plugin.

In the worst case, i.e., when the buffer terminates before the output plugin finishes `execute_chunking`, we lose the data.
Since we can expect almost no chance of that occurring, it would be better to make that a limitation.

Reply:
Fixed.

Reply:
> In the worst case, i.e., when the buffer terminates before the output plugin finishes execute_chunking, we lose the data.
Since we can expect almost no chance of that occurring, it would be better to make that a limitation.

If this occurs, the following warning log occurs.

```
SourceOnlyBufferAgent: emit transaction failed: error_class=NoMethodError error="undefined method `get' for nil:NilClass" location=...
```